tests/test_builder.py

```diff
@@ -0,0 +1,424 @@
+            GenericInstr.LOAD,
+            GenericInstr.BEQ,
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            PatternWildcard.BRANCH_LABEL,
+            GenericInstr.LOAD,
+            GenericInstr.BNZ,
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            PatternWildcard.BRANCH_LABEL,
+        ]
+    )
+
+
+def test_nested():
+    with DebugConnection("Alice") as conn:
+
+        q = Qubit(conn)
+        m = q.measure()
+        with m.if_eq(0):
+            with m.if_eq(1):
+                _ = Qubit(conn)
+
+        with conn.loop(2):
+            with conn.loop(3):
+                _ = Qubit(conn)
+
+        with conn.loop(2):
+            with m.if_eq(0):
+                _ = Qubit(conn)
+
+        with m.if_eq(0):
+            with conn.loop(2):
+                _ = Qubit(conn)
+
+        subroutine = conn.builder.subrt_pop_pending_subroutine()
+        print(subroutine)
+
+    inspector = PreSubroutineInspector(subroutine)
+
+    assert inspector.match_pattern(
+        [
+            GenericInstr.BNE,
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            GenericInstr.BNE,
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            PatternWildcard.BRANCH_LABEL,  # IF_EXIT
+            PatternWildcard.BRANCH_LABEL,  # IF_EXIT1
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            PatternWildcard.BRANCH_LABEL,  # LOOP1
+            GenericInstr.BEQ,
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            PatternWildcard.BRANCH_LABEL,  # LOOP
+            GenericInstr.BEQ,
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            GenericInstr.JMP,
+            PatternWildcard.BRANCH_LABEL,  # LOOP_EXIT
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            GenericInstr.JMP,
+            PatternWildcard.BRANCH_LABEL,  # LOOP_EXIT1
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            PatternWildcard.BRANCH_LABEL,  # LOOP2
+            GenericInstr.BEQ,
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            GenericInstr.BNE,
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            PatternWildcard.BRANCH_LABEL,  # IF_EXIT2
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            GenericInstr.JMP,
+            PatternWildcard.BRANCH_LABEL,  # LOOP_EXIT2
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            GenericInstr.BNE,
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            PatternWildcard.BRANCH_LABEL,  # LOOP3
+            GenericInstr.BEQ,
+            PatternWildcard.ANY_ZERO_OR_MORE,
+            GenericInstr.JMP,
+            PatternWildcard.BRANCH_LABEL,  # LOOP_EXIT3
+            PatternWildcard.BRANCH_LABEL,  # IF_EXIT3
+        ]
+    )
+
+
+def test_try():
+    with DebugConnection("Alice") as conn:
+
+        with conn.try_until_success(max_tries=1):
+            q = Qubit(conn)
+            q.measure()
+
+        subroutine = conn.builder.subrt_pop_pending_subroutine()
+        print(subroutine)
+
+
+if __name__ == "__main__":
+    # set_log_level("DEBUG")
```

Comment:
why not run all tests? As commented earlier why not run all the tests from the pipeline (only successful when all tests succeed).

Reply:
The test function calls in `if __name__ == "__main__"` are purely for the manual testing and debugging of a specific test  case. For example to debug one specific test case, it's easiest to comment out all other test case function calls, running `python tests/test_builder.py` (not `pytest`) and attaching a debugger. Which parts are commented out exactly is an artefact of manual testing/debugging. I uncommented everything again. To properly run the tests (also in the pipeline), `pytest` should be used.